packages/flutter/lib/src/material/scaffold.dart

```diff
@@ -61,6 +61,139 @@ enum _ScaffoldSlot {
+    @required this.child,
+  }) : assert(child != null),
+      super(key: key);
+
+  /// Returns the closest [ScaffoldMessengerState] which encloses the given
+  /// context.
+  ///
+  /// Typical usage is as follows:
+  ///
+  /// ```dart
+  /// ScaffoldMessages scaffoldMessenger = ScaffoldMessages.of(context);
+  /// scaffoldMessenger.showSnackBar(const SnackBar(content: Text('Snack-tastic'));
+  /// ```
+  static ScaffoldMessengerState of(BuildContext context) {
+    final _ScaffoldMessengerScope scope = context.dependOnInheritedWidgetOfExactType<_ScaffoldMessengerScope>();
+    return scope?._scaffoldMessengerState;
+  }
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  @override
+  ScaffoldMessengerState createState() => ScaffoldMessengerState();
+}
+
+/// Doc
+class ScaffoldMessengerState extends State<ScaffoldMessenger> {
+  final Set<ScaffoldState> _scaffolds = <ScaffoldState>{};
+  final Queue<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>> _snackBars = Queue<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>>();
+
+  void _register(ScaffoldState scaffold) {
+    // Are we in the middle of showing a SnackBar elsewhere?
+    // If so, present on the new scaffold.
+    if (!_scaffolds.contains(scaffold) && _snackBars.isNotEmpty) {
+      scaffold.showSnackBar(_snackBars.first._widget, listener: _handleSnackBarStatusChange);
+    }
+    _scaffolds.add(scaffold);
+  }
+
+  void _unregister(ScaffoldState scaffold) {
+    _scaffolds.remove(scaffold);
+  }
+
+  void _handleSnackBarStatusChange(AnimationStatus status) {
+    switch (status) {
+      case AnimationStatus.dismissed:
+        if (_snackBars.isNotEmpty)
+          _snackBars.removeFirst();
+        // If there is another Scaffold in the stack presenting a SnackBar, we
+        // want to make sure it has been dismissed too.
+        hideCurrentSnackBar();
+        break;
+      case AnimationStatus.completed:
+      case AnimationStatus.forward:
+      case AnimationStatus.reverse:
+        break;
+    }
+  }
+
+  /// Doc
+  ScaffoldFeatureController<SnackBar, SnackBarClosedReason> showSnackBar(SnackBar snackbar) {
+//    final Set<ScaffoldFeatureController<SnackBar, SnackBarClosedReason>> _controllers = <ScaffoldFeatureController<SnackBar, SnackBarClosedReason>>{};
+    for (final ScaffoldState scaffold in _scaffolds)
+      _snackBars.add(scaffold?.showSnackBar(snackbar, listener: _handleSnackBarStatusChange));
+
+    // TODO(Piinks): make sure we are returning the right controller since there
+    //  could be multiple across Scaffolds and one could have received input.
+    //  - We should track the snack bars differently:
+    //    - one object representing the SnackBars ScaffoldMessenger is currently managing
+    //    - one object representing the ScaffoldFeatureControllers generated for each SnackBar
+    //      - This will be the one that needs to return the right controller
```

Comment:
I don't fully understand this comment, so maybe this is what you already mean...

Shouldn't this return a forwarding ScaffoldFeatureController that basically forwards all actions a user can perform on it to the actual ScaffoldFeatureControllers of the individual snackbars?